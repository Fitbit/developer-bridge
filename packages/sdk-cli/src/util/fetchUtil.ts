import { decode } from '@fitbit/jsonrpc-ts';
import * as t from 'io-ts';

export const assertContentType = (expected: string) => (response: Response) => {
  // Parameters of media types are ignored.
  const contentTypeHeader = response.headers.get('content-type') || '';
  const contentType = contentTypeHeader.split(';', 1)[0];
  if (contentType !== expected) {
    return Promise.reject(
      new Error(`Unexpected Content-Type: expected '${expected}', got '${contentType}'`),
    );
  }
  return Promise.resolve(response);
};

export const assertJSON = (expected = 'application/json') => async (response: Response) => {
  await assertContentType(expected)(response);
  return response.json();
};

export const decodeJSON = <A, O, I>(endpointType: t.Type<A, O, I>) =>
  (response: Response): Promise<A> =>
    assertJSON()(response)
      .then(decode(endpointType));
