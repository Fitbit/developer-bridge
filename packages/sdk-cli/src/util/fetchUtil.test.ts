import { Response } from '../fetch';
import * as fetchUtil from './fetchUtil';

function makeResponse(
  init: ResponseInit = { status: 200 },
  body = '{}',
) {
  const response = new Response(
    body,
    {
      statusText: `Status ${init.status}`,
      ...init,
    },
  );

  Object.defineProperty(response, 'url', {
    value: 'http://api',
    writable: false,
  });

  return response;
}

describe('assertJSON()', () => {
  it.each([
    'application/text',
    'image/png',
    undefined,
  ])('rejects on Content-Type header of %s', (contentType) => {
    return expect(
      fetchUtil.assertJSON()(
        makeResponse({ headers: { 'Content-Type': contentType } }),
      ),
    ).rejects.toThrowErrorMatchingSnapshot();
  });

  it.each([
    'application/json',
    'application/json; charset=utf-8',
  ])('resolves on Content-Type header of %s', (contentType) => {
    return expect(
      fetchUtil.assertJSON()(
        makeResponse({ headers: { 'Content-Type': contentType } }),
      ),
    ).resolves.toEqual({});
  });

  it('rejects if the response is not valid JSON', () => {
    return expect(
      fetchUtil.assertJSON()(
        makeResponse(
          { headers: { 'Content-Type': 'application/json' } },
          '{notvalid',
        ),
      ),
    ).rejects.toThrowErrorMatchingSnapshot();
  });

  it('accepts an alternative Content-Type to assert', () => {
    return expect(
      fetchUtil.assertJSON('application/vnd.document+json')(
        makeResponse({ headers: { 'Content-Type': 'application/vnd.document+json' } }),
      ),
    ).resolves.toEqual({});
  });
});
