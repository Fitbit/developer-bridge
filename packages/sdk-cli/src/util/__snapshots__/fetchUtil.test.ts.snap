// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`assertJSON() rejects if the response is not valid JSON 1`] = `"invalid json response body at http://api reason: Unexpected token n in JSON at position 1"`;

exports[`assertJSON() rejects on Content-Type header of application/text 1`] = `"Unexpected Content-Type: expected 'application/json', got 'application/text'"`;

exports[`assertJSON() rejects on Content-Type header of image/png 1`] = `"Unexpected Content-Type: expected 'application/json', got 'image/png'"`;

exports[`assertJSON() rejects on Content-Type header of undefined 1`] = `"Unexpected Content-Type: expected 'application/json', got 'undefined'"`;
